network/crl_test.go

```diff
@@ -0,0 +1,536 @@
+			nextCert = verifiedChains[0][i+1]
+		}
+
+		if nextCert != nil {
+			// Every certificate (except last) should be signed by the next one
+			err := cert.CheckSignatureFrom(nextCert)
+			if err != nil {
+				t.Fatalf("Certificate `%s` was not signed by `%s`: %v\n", cert.Subject.String(), nextCert.Subject.String(), err)
+			}
+		}
+	}
+
+	return rawCerts, verifiedChains
+}
+
+func getValidTestChain(t *testing.T) ([][]byte, [][]*x509.Certificate) {
+	return getTestChain(
+		t,
+		TestCertPrefix,
+		TestCertFilename,
+		TestCACertFilename)
+}
+
+func getValidTestChain3(t *testing.T) ([][]byte, [][]*x509.Certificate) {
+	return getTestChain(
+		t,
+		TestCertPrefix3,
+		TestCertFilename3,
+		TestICACertFilename3,
+		TestCACertFilename3)
+}
+
+func getInvalidTestChain(t *testing.T) ([][]byte, [][]*x509.Certificate) {
+	return getTestChain(
+		t,
+		TestCertPrefix,
+		TestRevokedCertFilename,
+		TestCACertFilename)
+}
+
+func getInvalidTestChain3(t *testing.T) ([][]byte, [][]*x509.Certificate) {
+	return getTestChain(
+		t,
+		TestCertPrefix3,
+		TestRevokedCertFilename3,
+		TestICACertFilename3,
+		TestCACertFilename3)
+}
+
+func TestCRLConfig(t *testing.T) {
+	expectOk := func(url, fromCert string, checkWholeChain bool, cacheSize, cacheTime uint) {
+		config, err := NewCRLConfig(url, fromCert, checkWholeChain, cacheSize, cacheTime)
+		if config == nil || err != nil {
+			t.Logf("url=%v, fromCert=%v, checkWholeChain=%v, cacheSize=%v, cacheTime=%v\n",
+				url, fromCert, checkWholeChain, cacheSize, cacheTime)
+			t.Logf("config=%v, err=%v\n", config, err)
+			t.Fatal("Got `nil` result or unexpected error")
+		}
+	}
+
+	expectErr := func(url, fromCert string, checkWholeChain bool, cacheSize, cacheTime uint) {
+		config, err := NewCRLConfig(url, fromCert, checkWholeChain, cacheSize, cacheTime)
+		if config != nil || err == nil {
+			t.Logf("url=%v, fromCert=%v, checkWholeChain=%v, cacheSize=%v, cacheTime=%v\n",
+				url, fromCert, checkWholeChain, cacheSize, cacheTime)
+			t.Logf("config=%v, err=%v\n", config, err)
+			t.Fatal("Got unexpected result or `nil` error")
+		}
+	}
+
+	// Here we test valid and invalid URLs, all possible values for
+	// `--tls_crl_from_cert`, and some invalid ones too,
+	// valid and invalid uint values for `--tls_crl_cache_size`
+
+	expectOk("", CrlFromCertUseStr, false, 1, 5)
+	expectOk("", CrlFromCertIgnoreStr, false, 2, 5)
+	expectOk("http://127.0.0.1/main_crl.pem", CrlFromCertUseStr, false, 1, 5)
+	expectOk("", CrlFromCertUseStr, false, 100, 0)
+	expectOk("", CrlFromCertUseStr, true, 0, 1)
+	expectOk("", CrlFromCertUseStr, false, 1, 300)
+	expectOk("", CrlFromCertTrustStr, true, 1, 0)
+	expectOk("", CrlFromCertPreferStr, false, 1, 0)
+
+	expectErr("htt://invalid url", CrlFromCertUseStr, false, 1, 5)
+	expectErr("", "IgNoRe", false, 1, 5)
+	expectErr("", CrlFromCertUseStr, false, 2147483648, 5)
+	expectErr("", CrlFromCertUseStr, false, 1, 301)
+	expectErr("", CrlFromCertUseStr, false, 1, 9000)
```

Comment:
I'd add a comment at the end of each line with the reason why these values should be rejected.

Reply:
Done